modal/runner/shared/quantize.py

```diff
@@ -0,0 +1,155 @@
+    .pip_install("hf-transfer==0.1.4")
+    .pip_install("transformers==4.31.0")
+    .env(
+        {
+            "HF_HUB_ENABLE_HF_TRANSFER": "1",
+            "HF_HUB_DISABLE_PROGRESS_BARS": "1",
+        }
+    )
+)
+
+
+@stub.function(
+    image=quantizer_image,
+    gpu=modal.gpu.H100(count=1),
+    volumes={models_path: models_volume},
+    secrets=[Secret.from_name("huggingface"), *get_observability_secrets()],
+    timeout=3600 * 3,  # 3 hours
+)
+def quantize_model(
+    pretrained: str = "sophosympatheia/Midnight-Rose-70B-v2.0.3",
+    quantized: str = "sambarnes/Midnight-Rose-70B-v2.0.3-GPTQ",
+):
+    from auto_gptq import AutoGPTQForCausalLM, BaseQuantizeConfig
+    from transformers import AutoTokenizer
+
+    pretrained_model_path = get_model_path(pretrained)
+    quantized_model_path = get_model_path(quantized)
+    if not does_model_exist(pretrained_model_path):
+        logger.error("Pretrained does not exist at", pretrained_model_path)
+        return
+    elif does_model_exist(quantized_model_path):
+        logger.info("Quantized model already exists at", quantized_model_path)
+        return
+
+    logger.info("Loading calibration dataset...")
+    tokenizer = AutoTokenizer.from_pretrained(
+        pretrained_model_path,
+        use_fast=True,
+    )
+    examples = load_open_instruct(tokenizer=tokenizer)
+    examples_for_quant = [
+        {
+            "input_ids": example["input_ids"],
+            "attention_mask": example["attention_mask"],
+        }
+        for example in examples
+    ]
+
+    # By default, the model will always be loaded into CPU memory
+    logger.info("Loading pretrained model into memory...")
+    model = AutoGPTQForCausalLM.from_pretrained(
+        pretrained_model_path,
+        quantize_config=BaseQuantizeConfig(
+            bits=4,
+            desc_act=False,  # False can significantly speed up inference but might slightly decrease perplexity
+            damp_percent=0.1,  # Affects how samples are processed, 0.1 results in slightly better accuracy
+        ),
+    )
+
+    # Examples are dict whose keys can only be "input_ids" and "attention_mask".
+    # This is called the "GPTQ dataset" in TheBloke's READMEs, where he says:
+    #
+    #   "The calibration dataset used during quantisation. Using a dataset more
+    #    appropriate to the model's training can improve quantisation accuracy.
+    #    Note that the GPTQ calibration dataset is not the same as the dataset
+    #    used to train the model - please refer to the original model repo for
+    #    details of the training dataset(s)."
+
+    logger.info("Starting quantization...")
+    model.quantize(examples_for_quant)
+    model.save_quantized(quantized_model_path, use_safetensors=True)
+
+    logger.info("Committing to volume...")
+    models_volume.commit()
+    logger.info(f"Volume now contains {quantized_model_path}")
+
+
+def load_open_instruct(tokenizer, n_samples=128):
```

Comment:
this calibration dataset loading is mostly forked from the examples here:
https://github.com/AutoGPTQ/AutoGPTQ/blob/main/examples/quantization/quant_with_alpaca.py

where instead of reading from a json file, we sample from this dataset instead:
https://huggingface.co/datasets/VMware/open-instruct

only adjustments to the code were to limit number of samples & change column labels for the dataset